src/extended_range_proof.rs

```diff
@@ -0,0 +1,162 @@
+        statements: Vec<&ExtendedStatement<Self::PK>>,
+    ) -> Result<Vec<Option<ExtendedMask<Self::K>>>, RangeProofError>;
+
+    /// Verify the batch of range proofs against the given commitments and optional minimum value promises. If this
+    /// function returns Ok, it attests to the batch of commitments having values in the range [min_val_promise; 2^64-1]
+    /// and that the provers knew both the values and private keys for those commitments.
+    fn verify_batch(
+        &self,
+        proofs: Vec<&Self::Proof>,
+        statements: Vec<&ExtendedStatement<Self::PK>>,
+    ) -> Result<(), RangeProofError>;
+
+    /// Recover the (unverified) mask for a non-aggregated proof using the provided seed-nonce.
+    fn recover_mask(
+        &self,
+        proof: &Self::Proof,
+        statement: &ExtendedStatement<Self::PK>,
+    ) -> Result<Option<ExtendedMask<Self::K>>, RangeProofError>;
+
+    /// Verify a recovered mask for a non-aggregated proof against the commitment.
+    fn verify_mask(
+        &self,
+        commitment: &HomomorphicCommitment<Self::PK>,
+        extended_mask: &ExtendedMask<Self::K>,
+        value: u64,
+    ) -> Result<bool, RangeProofError>;
+}
+
+/// Rewind data extracted from a range proof containing the mask (e.g. blinding factor).
+#[derive(Debug, Clone, PartialEq)]
+pub struct ExtendedMask<K>
+where K: SecretKey
+{
+    blindings: Vec<K>,
+}
+
+impl<K> ExtendedMask<K>
+where K: SecretKey
+{
+    /// Construct a new extended mask
+    pub fn assign(extension_degree: ExtensionDegree, blindings: Vec<K>) -> Result<ExtendedMask<K>, RangeProofError> {
+        if blindings.is_empty() || blindings.len() != extension_degree as usize {
+            Err(RangeProofError::InitializationError(
+                "Extended mask length must correspond to the extension degree".to_string(),
+            ))
+        } else {
+            Ok(Self { blindings })
+        }
+    }
+
+    /// Return the extended mask blinding factors
+    pub fn blindings(&self) -> Result<Vec<K>, RangeProofError> {
+        if self.blindings.is_empty() {
+            Err(RangeProofError::InitializationError(
+                "Extended mask values not assigned yet".to_string(),
+            ))
+        } else {
+            Ok(self.blindings.clone())
+        }
+    }
+}
+
+/// The range proof statement contains thevector of commitments, vector of optional minimum promised
+/// values and a vector of optional seed nonces for mask recovery
+#[derive(Clone)]
+pub struct ExtendedStatement<PK>
+where PK: PublicKey
+{
+    /// The aggregated commitments
+    pub commitments: Vec<HomomorphicCommitment<PK>>,
+    /// Optional minimum promised values
+    pub minimum_value_promises: Vec<Option<u64>>,
+    /// Optional seed nonce for mask recovery
+    pub seed_nonce: Option<PK::K>,
+}
+
+impl<PK> ExtendedStatement<PK>
+where PK: PublicKey
+{
+    /// Initialize a new 'ExtendedStatement' with sanity checks
+    pub fn init(
```

Comment:
Or even better, put them in one vector
`commitments: Vec<(Commitment<PK>, u64)>`